HARK/ConsumptionSaving/ConsRiskyContribModel.py

```diff
@@ -0,0 +1,1768 @@
+
+        # Each stage has its own states and controls, and its methods
+        # to find them.
+        self.set_states = {
+            "Reb": self.get_states_Reb,
+            "Sha": self.get_states_Sha,
+            "Cns": self.get_states_Cns,
+        }
+
+        self.set_controls = {
+            "Reb": self.get_controls_Reb,
+            "Sha": self.get_controls_Sha,
+            "Cns": self.get_controls_Cns,
+        }
+
+        # Set the solver for the portfolio model, and update various constructed attributes
+        self.solve_one_period = solveRiskyContrib
+        self.update()
+
+    def pre_solve(self):
+        self.update_solution_terminal()
+
+    def update(self):
+
+        RiskyAssetConsumerType.update(self)
+        self.update_share_grid()
+        self.update_dfrac_grid()
+        self.update_nNrm_grid()
+        self.update_mNrm_grid()
+        self.update_tau()
+
+    def update_solution_terminal(self):
+        """
+        Solves the terminal period. The solution is trivial.
+        Cns: agent will consume all of his liquid resources.
+        Sha: irrelevant as there is no "next" period.
+        Reb: agent will shift all of his resources to the risk-free asset.
+        
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        None
+        """
+
+        # Construct the terminal solution backwards.
+
+        # Start with the consumption stage. All liquid resources are consumed.
+        cFunc_term = IdentityFunction(i_dim=0, n_dims=3)
+        vFunc_Cns_term = ValueFuncCRRA(cFunc_term, CRRA=self.CRRA)
+        # Marginal values
+        dvdmFunc_Cns_term = MargValueFuncCRRA(cFunc_term, CRRA=self.CRRA)
+        dvdnFunc_Cns_term = ConstantFunction(0.0)
+        dvdsFunc_Cns_term = ConstantFunction(0.0)
+
+        Cns_stage_sol = RiskyContribCnsSolution(
+            # Consumption stage
+            vFunc_Cns=vFunc_Cns_term,
+            cFunc=cFunc_term,
+            dvdmFunc_Cns=dvdmFunc_Cns_term,
+            dvdnFunc_Cns=dvdnFunc_Cns_term,
+            dvdsFunc_Cns=dvdsFunc_Cns_term,
+        )
+
+        # Share stage
+
+        # It's irrelevant because there is no future period. Set share to 0.
+        # Create a dummy 2-d consumption function to get value function and marginal
+        c2d = IdentityFunction(i_dim=0, n_dims=2)
+        Sha_stage_sol = RiskyContribShaSolution(
+            # Adjust
+            vFunc_Sha_Adj=ValueFuncCRRA(c2d, CRRA=self.CRRA),
+            ShareFunc_Adj=ConstantFunction(0.0),
+            dvdmFunc_Sha_Adj=MargValueFuncCRRA(c2d, CRRA=self.CRRA),
+            dvdnFunc_Sha_Adj=ConstantFunction(0.0),
+            # Fixed
+            vFunc_Sha_Fxd=vFunc_Cns_term,
+            ShareFunc_Fxd=IdentityFunction(i_dim=2, n_dims=3),
+            dvdmFunc_Sha_Fxd=dvdmFunc_Cns_term,
+            dvdnFunc_Sha_Fxd=dvdnFunc_Cns_term,
+            dvdsFunc_Sha_Fxd=dvdsFunc_Cns_term,
+        )
+
+        # Rabalancing stage
```

Comment:
r_e_balancing

Reply:
https://github.com/econ-ark/HARK/pull/832/commits/cd2b3130c2f89bf6f1f9cc2e7109e3217de59273

Thank you!